gh-75988: Fix issues with autospec ignoring wrapped object (#115223)
python/cpython: Doc/library/unittest.mock.rst

        >>> mock.not_submock.attribute2  # This won't raise.

    .. versionadded:: 3.7

> Order of precedence of side_effect, return_value and wraps
> ----------------------------------------------------------

Reviewer: This would be better, providing it renders okay.
```suggestion
Order of precedence of :attr:`side_effect`, :attr:`return_value` and *wraps*
----------------------------------------------------------------------------
```

The order of their precedence is:

> :attr:`~Mock.side_effect` > :attr:`~Mock.return_value` > *wraps*

Reviewer: ```suggestion
1. :attr:`~Mock.side_effect`
2. :attr:`~Mock.return_value`
3. *wraps*
```

> If all three are set, mock will return the value from :attr:`~Mock.side_effect`

Reviewer: ```suggestion
If all three are set, mock will return the value from :attr:`~Mock.side_effect`,
```

    >>> order_mock.get_value.return_value = "second"
    >>> order_mock.get_value()
    'first'

As ``None`` is the default value of :attr:`~Mock.side_effect`, if you reassign
its value back to ``None``, the order of precedence will be checked between
:attr:`~Mock.return_value` and the wrapped object, ignoring
:attr:`~Mock.side_effect`.

    >>> order_mock.get_value.side_effect = None
    >>> order_mock.get_value()
    'second'

If the value being returned by :attr:`~Mock.side_effect` is :data:`DEFAULT`,
it is ignored and the order of precedence moves to the successor to obtain the
value to return.

    >>> from unittest.mock import DEFAULT
    >>> order_mock.get_value.side_effect = [DEFAULT]
    >>> order_mock.get_value()
    'second'

Unless explicitly set, the default value of :attr:`~Mock.return_value` is
:data:`DEFAULT`.

    >>> order_mock = Mock(spec=Order, wraps=Order)
    >>> order_mock.return_value
    sentinel.DEFAULT
    >>> order_mock.get_value.return_value
    sentinel.DEFAULT

Reassigning back to this value or not explicitly assigning any value to it,
the order of precedence will ignore it and move to the last successor which is

> the wrapped object.

Reviewer: Not sure this paragraph is necessary.

Author: I have removed it but with a slight change.

https://github.com/python/cpython/blob/49a08dc4805a4c0c6ec1095964b9dde89fdf6f8c/Doc/library/unittest.mock.rst#L2881-L2891

```diff
+ When :class:`Mock` wraps an object, the default value of 
+ :attr:`~Mock.return_value` will be :data:`DEFAULT`.
```

This informtion can be verified from this test case 👇:
https://github.com/python/cpython/blob/49a08dc4805a4c0c6ec1095964b9dde89fdf6f8c/Lib/test/test_unittest/testmock/testmock.py#L680-L691

> _It will not be "sentinel.DEFAULT" if the mock is not wrapping any object._

**_If you agree with the change, mark it as resolved._**

As the real call is being made to the wrapped object, creating an instance of

> this mock will return the real instance of the class. The positional arguments
> (if any) required by the wrapped object must be passed.

Reviewer: ```suggestion
this mock will return the real instance of the class. The positional arguments,
if any, required by the wrapped object must be passed.
```

    >>> isinstance(order_mock_instance, Order)
    True
    >>> order_mock_instance.get_value()
    'third'

    >>> order_mock.get_value.return_value = DEFAULT
    >>> order_mock.get_value()
    'third'

    >>> order_mock.get_value.return_value = "second"
    >>> order_mock.get_value()
    'second'

But if you assign ``None`` to it, this will not be ignored as it is an
explicit assignment. So, the order of precedence will not move to the wrapped
object.

    >>> order_mock.get_value.return_value = None
    >>> order_mock.get_value() is None
    True

Even if you set all three at once when initializing the mock, the order of
precedence remains the same:

    >>> order_mock = Mock(spec=Order, wraps=Order,
    ...                   **{"get_value.side_effect": ["first"],
    ...                      "get_value.return_value": "second"}
    ...                   )
    ...
    >>> order_mock.get_value()
    'first'
    >>> order_mock.get_value.side_effect = None
    >>> order_mock.get_value()
    'second'
    >>> order_mock.get_value.return_value = DEFAULT
    >>> order_mock.get_value()
    'third'

If :attr:`~Mock.side_effect` is exhausted, the order of precedence will not

> obtain value to return from the successors. Instead, ``StopIteration``

Reviewer: ```suggestion
cause a value to be obtained from the successors. Instead, ``StopIteration``
```
